Added context menu to editor

petith-au opened the pull request:
### Added context menu to editor

**Requirements:**
Add a right-click menu on the text-editor to allow for copy and paste functionality.

- Disable copy menu item if no text in the editor is selected.
- Disable paste menu item if no text is available in the clipboard.

**Not included:**
WordWrap requirement, as I see this being implemented with the full clipboard scope https://github.com/udlose/MermaidPad/issues/106

This PR implements the context menu, including such that full clipboard functionality can be added. 

Could definitely improve how we handle the clipboard service, but would like your input on this before I go further OR it could just be implemented completely with #106.

Either way let me know what changes you would like to see.

udlose commented on Views/MainWindow.axaml.cs:
```diff
@@ -563,4 +567,29 @@ private void OnCloseClick(object? sender, RoutedEventArgs e)
+        //Get Clipboard state
+        _vm.CanCopyClipboard = _vm.EditorSelectionLength > 0;
+        _vm.CanPasteClipboard = GetTextFromClipboardAsync(this).Result != String.Empty;
+    }
```
then add this method:
```
    /// <summary>
    /// Reads clipboard text asynchronously and updates the ViewModel's CanPasteClipboard on the UI thread.
    /// </summary>
    private async Task UpdateCanPasteClipboardAsync()
    {
        string? clipboardText = null;

        try
        {
            // Perform clipboard I/O off the UI context
            clipboardText = await GetTextFromClipboardAsync(this)
                .ConfigureAwait(false);
        }
        catch (Exception ex)
        {
            // Log and treat as no pasteable text
            SimpleLogger.LogError("Error reading clipboard text", ex);
        }

        bool canPaste = !string.IsNullOrWhiteSpace(clipboardText);

        // Marshal back to UI thread to update the ViewModel property
        await Dispatcher.UIThread.InvokeAsync(() => _vm.CanPasteClipboard = canPaste, DispatcherPriority.Normal);
    }
```

petith-au replied:
Apologies. Overlooked this in my PR!

Appreciate the recommended changes, this makes sense to me. I will add this in.

udlose commented on Views/MainWindow.axaml.cs:
```diff
@@ -563,4 +567,29 @@ private void OnCloseClick(object? sender, RoutedEventArgs e)
+        var response = await clipboard.TryGetTextAsync();
+
+        return response;
+    }
```
```
        IClipboard? clipboard = window.Clipboard;
        if (clipboard is null)
        {
            return null;
        }

        string? clipboardText = await clipboard.TryGetTextAsync();

        return clipboardText;
```

petith-au replied:
updated this as well.

udlose reviewed (changes requested):
Nice start! I like the fact that you decided to break it into little pieces rather than a much larger PR :)

Just some small suggestions on the async piece to avoid blocking, possible deadlocks.

petith-au commented on Views/MainWindow.axaml.cs:
```diff
@@ -563,4 +567,29 @@ private void OnCloseClick(object? sender, RoutedEventArgs e)
+    {
+        //Get Clipboard state
+        _vm.CanCopyClipboard = _vm.EditorSelectionLength > 0;
+        _vm.CanPasteClipboard = GetTextFromClipboardAsync(this).Result != String.Empty;
```
Awesome, thanks for this feedback.
Wasn't aware of the .SafeFireAndForget, so I've updated this now!

petith-au commented:
@udlose I've updated the PR.

I've removed the comments too, as I understood the changes you recommended! But let me know if you would prefer them back in, for future contributions.

petith-au commented on Views/MainWindow.axaml:
```diff
@@ -70,15 +70,27 @@
+						IsEnabled="{Binding CanCopyClipboard}"/>
+					<MenuItem 
+						Header="Paste" 
+						IsEnabled="{Binding CanPasteClipboard}"/>
```
@udlose I can add placeholder handlers, otherwise we can tackle this in the other ticket?

petith-au commented on Views/MainWindow.axaml.cs:
```diff
@@ -267,6 +269,8 @@ private void OnViewModelPropertyChanged(object? sender, PropertyChangedEventArgs
             case nameof(_vm.EditorSelectionStart):
             case nameof(_vm.EditorSelectionLength):
+            case nameof(_vm.CanCopyClipboard):
+            case nameof(_vm.CanPasteClipboard):
```
@udlose possibly don't need this, but I have added this as a placeholder. If this switch statement is just for the editor state then I am inclined to think we would only need this for the paste operation itself.

udlose commented on Views/MainWindow.axaml.cs:
```diff
@@ -563,4 +567,58 @@ private void OnCloseClick(object? sender, RoutedEventArgs e)
+        catch (Exception ex)
+        {
+            SimpleLogger.LogError("Error reading clipboard text", ex);
+        }
```
this is fine for now. you can change if u want or not. since this is still a 1.x version, I error on the safer side with having a "catch all".
